Ecal/src/Ecal/EcalWABRecProcessor.cxx

```diff
@@ -0,0 +1,847 @@
+  std::vector<float> recoil_e_pos, recoil_y_pos;
+
+  // Result object that stores kinematic variables
+  ldmx::EcalWABResult result;
+
+  // Define kinematic variables
+  const std::vector<double> z_hat = {0, 0, 1};
+  double true_theta_electron = -9.;
+  double true_theta_photon = -9.;
+  double true_phi_electron = -9.;
+  double true_phi_photon = -9.;
+  double rec_theta_electron = -9.;
+  double rec_theta_photon = -9.;
+  double rec_phi_electron = -9.;
+  double rec_phi_photon = -9.;
+  double true_theta_diff_electron_photon = -9.;
+  double true_phi_diff_electron_photon = -9.;
+  double rec_theta_diff_electron_photon = -9.;
+  double rec_phi_diff_electron_photon = -9.;
+  double true_rec_theta_diff_electron = -9.;
+  double true_rec_phi_diff_electron = -9.;
+  double true_rec_theta_diff_photon = -9.;
+  double true_rec_phi_diff_photon = -9.;
+  float true_electron_shower_energy = -999.;
+  float true_photon_shower_energy = -999.;
+  double rec_electron_shower_energy = -999.;
+  double rec_photon_shower_energy = -999.;
+
+  // Create lists for rec hits and electron/photon shower hits
+  std::vector<std::array<double, 6>> rec_hit_list;
+  std::vector<std::array<double, 6>> ele_hit_list;
+  std::vector<std::array<double, 6>> phot_hit_list;
+
+  // Save rec hit info to rec_hit_list
+  for (const ldmx::EcalHit& hit : ecal_rec_hits) {
+    ldmx::EcalID id(hit.getID());
+    auto [x, y, z] = geometry_->getPosition(id);
+    double energy = hit.getEnergy();
+    double layer_num = id.layer();
+    rec_hit_list.push_back({x, y, z, layer_num, 0, energy});
+  }
+
+  if (event.exists("TargetScoringPlaneHits")) {
+    //
+    // Loop through all of the sim particles and find the recoil
+    // photon/electron.
+    //
+
+    // Find Target SP hit for recoil photon/electron
+    const std::vector<ldmx::SimTrackerHit> target_sp_hits =
+        event.getCollection<ldmx::SimTrackerHit>("TargetScoringPlaneHits");
+    float photon_p_zmax = 0, electron_p_zmax = 0;
+    for (const ldmx::SimTrackerHit& sp_hit : target_sp_hits) {
+      ldmx::SimSpecialID hit_id(sp_hit.getID());
+      if (hit_id.plane() != 1 || sp_hit.getMomentum()[2] <= 0) continue;
+
+      if (sp_hit.getPdgID() == 11) {
+        if (sp_hit.getMomentum()[2] > electron_p_zmax) {
+          recoil_e_p = sp_hit.getMomentum();
+          true_electron_shower_energy = sp_hit.getEnergy();
+          recoil_e_pos = sp_hit.getPosition();
+          electron_p_zmax = recoil_e_p[2];
+        }
+      }
+      if (sp_hit.getPdgID() == 22) {
+        if (sp_hit.getMomentum()[2] > photon_p_zmax) {
+          recoil_y_p = sp_hit.getMomentum();
+          true_photon_shower_energy = sp_hit.getEnergy();
+          recoil_y_pos = sp_hit.getPosition();
+          photon_p_zmax = recoil_y_p[2];
+        }
+      }
+    }
+
+    // Calculating true theta values using SP hit parameters
+    if (recoil_y_p.size() == 3 &&
+        std::sqrt(std::pow(recoil_y_p[0], 2) + std::pow(recoil_y_p[1], 2) +
```

Comment:
the `std:pow` is actually pretty costly, so instead of `std::pow(x, 2` it's more efficient to do x*x

Reply:
if you move `recoil_y_p` to be an array then the size check is not necessary anymore